### DISCUSSION TITLE
Automatically adding query parameter to each route in OpenAPI specification

### DISCUSSION THREAD

**Discussion by ccharlesgb** [fastapi/fastapi#7189] [Questions] [answered]
### Example

Here is a small example of my use case, there is a legacy reason why this situation has occurred. Basically we are moving something that used to be a query parameter in all of our routes in a service to being automatically supplied in the header by the reverse proxy in front of the service. All of our routes used to have defined the `team` Query parameter but now this will be no longer used as it will be supplied in the header, I need the middleware to temporarily accept the header and the query parameter as a value as not all services will be updated at once:

```Python
from typing import Optional

from fastapi import FastAPI, APIRouter, Query, HTTPException
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.types import ASGIApp
from pydantic import BaseModel
from uvicorn import run

class UserCreds(BaseModel):
    team: Optional[str]

class MyAuthMiddleware(BaseHTTPMiddleware):
    """
    Middleware to inject the attribute `creds` into the request state
    """

    def __init__(self, app: ASGIApp):
        super().__init__(app)

    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint):
        query_param = request.query_params.get("team")
        header=request.headers.get("team")
        if query_param != header:
            raise HTTPException(400, "Query param and header don't align")
        request.state.creds = UserCreds(team=header)
        response = await call_next(request)
        return response


app = FastAPI(title="Test App")
app.add_middleware(MyAuthMiddleware)


router = APIRouter()

@router.get("/items/")
def get_items():
    return "Items"

# I'd like to not go through every route and keep this query parameter so it appears in OpenAPI docs
# Because of the middleware the query parameter is always used
@router.get("/things/")
def get_things(team: Optional[str] = Query(None)):
    return "Things"

app.include_router(router)

run(app)
```

### Description

Essentially what I am trying to do is make it so if a query parameter is used by the middleware then it should insert that query parameter into every route in the OpenAPI specification. I couldn't see anything in the docs that would achieve this. I know that even if I remove the query parameter in the route definition this code will still work but it would be nice if it appeared in the Swagger UI as well.

### Environment

* OS: Linux
* FastAPI Version: 0.61.1
* Python version: 3.8.5

**Comment by includeamin** [comment, marked as answer]
Hi, if I understand correctly you would like the `team` query to be applied to all routes on the `router` in the `Swagger UI`.
with dependencies on `include_router`, query appear on `Swagger UI`
``` python
def get_team_query(team:str):
   pass

@router.get("/sample" )
def sample_2(request: Request):
    return request.query_params.get('team')

@router.get("/item" )
def items(request: Request):
    return request.query_params.get('team')


@router.get("/sample")
def sample_1(request: Request):
    return request.query_params.get('team')

app.include_router(router,dependencies=[Depends(get_team_query)])

```

**Comment by ccharlesgb** [comment]
That will work for my use case yes! It would be interesting to be able to add OpenAPI parameters from middleware but that is maybe quite a niche feature.

**Comment by Mause** [comment]
If you problem is solved, can you please close this issue?

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue :+1:
